core/trino-main/src/main/java/io/trino/operator/aggregation/DistinctWindowAccumulator.java

```diff
@@ -0,0 +1,195 @@
+                argumentTypes,
+                argumentChannels,
+                hash.copy(),
+                pagesIndexFactory);
+    }
+
+    @Override
+    public void addInput(WindowIndex index, int startPosition, int endPosition)
+    {
+        // index is remapped so just go from 0 to argumentChannels.size()
+        for (int position = startPosition; position <= endPosition; position++) {
+            if (pageBuilder.isFull()) {
+                indexCurrentPage();
+            }
+            for (int channel = 0; channel < argumentChannels.size(); channel++) {
+                ValueBlock value = index.getSingleValueBlock(channel, position).getSingleValueBlock(0);
+                pageBuilder.getBlockBuilder(channel).append(value, 0);
+            }
+            pageBuilder.declarePosition();
+        }
+    }
+
+    private void indexCurrentPage()
+    {
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
+        Work<Block> work = hash.markDistinctRows(page);
+        checkState(work.process());
+        Block distinctMask = work.getResult();
+
+        int positionCount = distinctMask.getPositionCount();
+        checkArgument(positionCount == page.getPositionCount(), "Block position count does not match current position count");
+        PagesIndex pagesIndex = pagesIndexFactory.newPagesIndex(argumentTypes, positionCount);
+
+        // short circuit if there are no nulls, and we are not testing the value
+        if (distinctMask instanceof RunLengthEncodedBlock) {
+            if (test(distinctMask, 0)) {
+                // all positions selecte
+                pagesIndex.addPage(page);
+            }
+            // no positions selected
+            return;
+        }
+        else {
+            int selectedPositionsCount = 0;
+            int[] selectedPositions = new int[positionCount];
+            for (int position = 0; position < positionCount; position++) {
+                if (test(distinctMask, position)) {
+                    selectedPositions[selectedPositionsCount] = position;
+                    selectedPositionsCount++;
+                }
+            }
+
+            PageBuilder filteredPageBuilder = new PageBuilder(argumentTypes);
+            for (int channel = 0; channel < argumentChannels.size(); channel++) {
+                BlockBuilder filteredBlockBuilder = filteredPageBuilder.getBlockBuilder(channel);
+                Block block = page.getBlock(channel);
+                for (int index = 0; index < selectedPositionsCount; index++) {
+                    int position = selectedPositions[index];
+                    // todo make it more efficient
+                    ValueBlock singleValueBlock = block.getSingleValueBlock(position);
+                    filteredBlockBuilder.append(singleValueBlock, 0);
+                }
+            }
+            filteredPageBuilder.declarePositions(selectedPositionsCount);
+            pagesIndex.addPage(filteredPageBuilder.build());
```

Comment:
I think this code would be easier to follow if we didn't materialize `selectedPositions[]` and instead just appended values for the current selected position in a nested loop.

Reply:
Yeah - restructured a bit.
Also added handling of ValueBlock so we do not have to call getSingleValueBlock if  not needed